REGRESSION(257861@main?): 8X Scrolling API-Tests are constantly failing or timing out

whsieh opened the pull request:
#### 3f19d05a7336ccd24aa222e8a132cd3fc2a119dd
<pre>
REGRESSION(257861@main?): 8X Scrolling API-Tests are constantly failing or timing out
<a href="https://bugs.webkit.org/show_bug.cgi?id=249339">https://bugs.webkit.org/show_bug.cgi?id=249339</a>
rdar://103373219

Reviewed by Aditya Keerthi.

Some of these API tests began failing because we&apos;re attempting to wait for `KeyboardWillShow`
notifications in the case where the keyboard never actually shows up (due to the fact that API tests
run in a non-UI-app context). Mitigate this by checking for the existence of an `UIKeyboard` right
after showing the keyboard, after focusing an editable element.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _didFinishTextInteractionInTextInputContext:]):

Additionally, some of the `RequestTextInputContext` API tests are legitimately failing due to the
fact that they expect us to zoom to reveal the focused element after finishing interaction in a
`_WKTextInputContext`; the way this worked before is that we set the
`m_waitingForPostLayoutEditorStateUpdateAfterFocusingElement` flag, which would cause us to then
call into `-_zoomToRevealFocusedElement` after the next editor state update.

To fix this, we create a new `_revealFocusedElementDeferrer` in
`-_didFinishTextInteractionInTextInputContext:`, allowing us to reveal the focused element after the
next editor state arrives.

(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

Canonical link: <a href="https://commits.webkit.org/257893@main">https://commits.webkit.org/257893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6cb2b28e3185c28fc7a96acd7249888658fa26a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100328 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33393 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109647 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/28 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107525 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106106 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/92730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/92730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/3233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3225 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43536 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/5423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/5042 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2803 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -7049,6 +7050,13 @@ - (void)_elementDidFocus:(const WebKit::FocusedElementInformation&)information u
             [self.window makeKeyWindow];
     }
 
+    if (!UIKeyboard.activeKeyboard) {
```
Just to confirm, `UIKeyboard.activeKeyboard` does not imply anything about visibility state right?

(i.e. the keyboard can be hidden but `UIKeyboard.activeKeyboard` is non-nil)

whsieh replied:
Yeah, that's correct — though, the intent was really to detect this "non-UI application" scenario and avoid waiting for the keyboardWillShow notification that will never arrive.

I'll poke around a bit more and see if there's a cleaner way to achieve this...

pxlcoder replied:
I think this is fine – was just concerned given the `active` in the name of the method.

whsieh committed:
REGRESSION(257861@main?): 8X Scrolling API-Tests are constantly failing or timing out
https://bugs.webkit.org/show_bug.cgi?id=249339
rdar://103373219

Reviewed by Aditya Keerthi.

Some of these API tests began failing because we're attempting to wait for `KeyboardWillShow`
notifications in the case where the keyboard never actually shows up (due to the fact that API tests
run in a non-UI-app context). Mitigate this by checking for the existence of an `UIKeyboard` right
after showing the keyboard, after focusing an editable element.

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _didFinishTextInteractionInTextInputContext:]):

Additionally, some of the `RequestTextInputContext` API tests are legitimately failing due to the
fact that they expect us to zoom to reveal the focused element after finishing interaction in a
`_WKTextInputContext`; the way this worked before is that we set the
`m_waitingForPostLayoutEditorStateUpdateAfterFocusingElement` flag, which would cause us to then
call into `-_zoomToRevealFocusedElement` after the next editor state update.

To fix this, we create a new `_revealFocusedElementDeferrer` in
`-_didFinishTextInteractionInTextInputContext:`, allowing us to reveal the focused element after the
next editor state arrives.

(-[WKContentView _elementDidFocus:userIsInteracting:blurPreviousNode:activityStateChanges:userObject:]):

Canonical link: https://commits.webkit.org/257893@main
